File: src/converter/plugins/NullableUnionTypePlugin.ts
import ts, {Node, Type, TypeNode, UnionTypeNode} from "typescript";
import {ConverterPlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {ConverterContext} from "../context.js";
import {Render, renderResolvedNullable} from "../render.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {resolveParenthesizedType} from "../../utils/resolveParenthesizedType.js";
import {isNullableStringUnionType} from "./StringUnionTypePlugin.js";
import {GeneratedFile} from "../generated.js";

const isNull = (type: Node) => ts.isLiteralTypeNode(type) && type.literal.kind === ts.SyntaxKind.NullKeyword
const isUndefined = (type: Node) => type.kind === ts.SyntaxKind.UndefinedKeyword

export const isNullableType = (type: Node) => isNull(type) || isUndefined(type)

export function isPossiblyNullableType(node: TypeNode, context: ConverterContext): boolean {
    if (node.getSourceFile() === undefined) {
        // handle synthetic nodes
        return false
    }

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    const typeChecker = typeScriptService?.program.getTypeChecker()

    const type = typeChecker?.getTypeAtLocation(node)

    if (!type) return false

    return (
        isNullableTsType(type)
        || (
            type.isUnion()
            && type.types.some(it => isNullableTsType(it))
        )
    )
}

function isNullableTsType(type: Type): boolean {
    return (
        (type.flags & ts.TypeFlags.Null) !== 0
        || (type.flags & ts.TypeFlags.Undefined) !== 0
        || (type.flags & ts.TypeFlags.Any) !== 0
        || (type.flags & ts.TypeFlags.Unknown) !== 0
    )
}

export function isNullableUnionType(node: Node, context: ConverterContext): node is UnionTypeNode {
    if (!ts.isUnionTypeNode(node)) return false

    return flatUnionTypes(node, context).some(type => isNullableType(type))
}

export function isNullableOnlyUnionType(node: Node, context: ConverterContext): node is UnionTypeNode {
    if (!ts.isUnionTypeNode(node)) return false

    return flatUnionTypes(node, context).every(type => isNullableType(type))
}

export function flatUnionTypes(node: UnionTypeNode, context: ConverterContext) {
    let result: TypeNode[] = []

    for (const type of node.types) {
        if (ts.isIndexedAccessTypeNode(type)) {
            const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

            const resolvedType = typeScriptService?.resolveType(type)

            if (resolvedType && ts.isUnionTypeNode(resolvedType)) {
                result = result.concat(flatUnionTypes(resolvedType, context))
            } else {
                result.push(type)
            }
        } else if (ts.isParenthesizedTypeNode(type)) {
            const resolvedType = resolveParenthesizedType(type)

            if (ts.isUnionTypeNode(resolvedType)) {
                result = result.concat(flatUnionTypes(resolvedType, context))
            } else {
                result.push(resolvedType)
            }
        } else {
            result.push(type)
        }
    }

    return result
}

export class NullableUnionTypePlugin implements ConverterPlugin {
    generate(): GeneratedFile[] {
        return [];
    }

    render(node: ts.Node, context: ConverterContext, next: Render): string | null {
        const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
        checkCoverageService?.cover(node)

        if (isNullableOnlyUnionType(node, context)) {
            node.types.forEach(it => checkCoverageService?.deepCover(it))

            return "Nothing?"
        }

        if (isNullableStringUnionType(node, context)) return null

        if (isNullableUnionType(node, context)) {
            const types = flatUnionTypes(node, context)

            const nonNullableTypes = types.filter(type => !isNullableType(type))
            const nullableTypes = node.types.filter(type => isNullableType(type))

            nullableTypes.forEach(it => checkCoverageService?.deepCover(it))

            if (nonNullableTypes.length === 1) {
                const nonNullableType = nonNullableTypes[0]

                // any and unknown case
                if (isPossiblyNullableType(nonNullableType, context)) {
                    return next(nonNullableType)
                }

                return renderResolvedNullable(nonNullableType, true, next)
            }
        }

        return null;
    }

    setup(context: ConverterContext): void {
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/NullableUnionTypePlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import io.github.sgrishchenko.karakum.util.resolveParenthesizedType
import js.array.ReadonlyArray
import typescript.*
import kotlin.contracts.contract

private fun isNull(type: Node) = isLiteralTypeNode(type) && type.literal.kind === SyntaxKind.NullKeyword
private fun isUndefined(type: Node) = type.kind === SyntaxKind.UndefinedKeyword
private fun isAny(type: Node) = type.kind === SyntaxKind.AnyKeyword
private fun isUnknown(type: Node) = type.kind === SyntaxKind.UnknownKeyword

fun isNullableType(type: Node) = isNull(type) || isUndefined(type)

@Suppress("UNCHECKED_CAST_TO_EXTERNAL_INTERFACE")
fun isPossiblyNullableType(node: TypeNode, context: Context): Boolean {
    if (node.getSourceFileOrNull() == null) {
        // handle synthetic nodes
        val resolvedType = resolveTypeIfNeeded(node, context)

        return (
            isNullableTsNode(resolvedType)
            || (
                isUnionTypeNode(resolvedType)
                && flatUnionTypes(resolvedType, context).any { isNullableTsNode(it) }
            )
        )
    }

    val typeScriptService = context.lookupService(typeScriptServiceKey)

    val typeChecker = typeScriptService?.program?.getTypeChecker()

    val type = typeChecker?.getTypeAtLocation(node) ?: return false

    return (
        isNullableTsType(type)
        || (
            type.isUnion()
            && (type as UnionType).types.any { isNullableTsType(it) }
        )
    )
}

private fun resolveTypeIfNeeded(node: TypeNode, context: Context): Node {
    val typeScriptService = context.lookupService(typeScriptServiceKey)

    if (isIndexedAccessTypeNode(node)) {
        val resolvedType = typeScriptService?.resolveType(node)

        return resolvedType ?: node
    }

    return node
}

private fun isNullableTsNode(node: Node): Boolean {
    return (
        isNull(node)
        || isUndefined(node)
        || isAny(node)
        || isUnknown(node)
    )
}

private fun isNullableTsType(type: Type): Boolean {
    // TODO: provide bit mask for TypeFlags
    return (
        (type.flags.toString().toInt() and TypeFlags.Null.toString().toInt()) != 0
        || (type.flags.toString().toInt() and TypeFlags.Undefined.toString().toInt()) != 0
        || (type.flags.toString().toInt() and TypeFlags.Any.toString().toInt()) != 0
        || (type.flags.toString().toInt() and TypeFlags.Unknown.toString().toInt()) != 0
    )
}

@Suppress("CANNOT_CHECK_FOR_EXTERNAL_INTERFACE")
fun isNullableUnionType(node: Node, context: Context): Boolean {
    contract {
        returns(true) implies (node is UnionTypeNode)
    }

    if (!isUnionTypeNode(node)) return false

    return flatUnionTypes(node, context).any { isNullableType(it) }
}

@Suppress("CANNOT_CHECK_FOR_EXTERNAL_INTERFACE")
fun isNullableOnlyUnionType(node: Node, context: Context): Boolean {
    contract {
        returns(true) implies (node is UnionTypeNode)
    }

    if (!isUnionTypeNode(node)) return false

    return flatUnionTypes(node, context).all { isNullableType(it) }
}

fun flatUnionTypes(node: UnionTypeNode, context: Context): ReadonlyArray<TypeNode> {
    val result = mutableListOf<TypeNode>()

    for (type in node.types) {
        if (isIndexedAccessTypeNode(type)) {
            val typeScriptService = context.lookupService(typeScriptServiceKey)

            val resolvedType = typeScriptService?.resolveType(type)

            if (resolvedType != null && isUnionTypeNode(resolvedType)) {
                result += flatUnionTypes(resolvedType, context).toList()
            } else {
                result += type
            }
        } else if (isParenthesizedTypeNode(type)) {
            val resolvedType = resolveParenthesizedType(type)

            if (isUnionTypeNode(resolvedType)) {
                result += flatUnionTypes(resolvedType, context).toList()
            } else {
                result += resolvedType
            }
        } else {
            result += type
        }
    }

    return result.toTypedArray()
}

class NullableUnionTypePlugin : Plugin {
    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()

    override fun render(node: Node, context: Context, next: Render<Node>): String? {
        val checkCoverageService = context.lookupService(checkCoverageServiceKey)
        checkCoverageService?.cover(node)

        if (isNullableOnlyUnionType(node, context)) {
            node.types.asArray().forEach { checkCoverageService?.deepCover(it) }

            return "Nothing?"
        }

        if (isNullableLiteralUnionType(node, context)) return null

        if (isNullableUnionType(node, context)) {
            val types = flatUnionTypes(node, context)

            val nonNullableTypes = types.filter{ !isNullableType(it) }
            val nullableTypes = node.types.asArray().filter { isNullableType(it) }

            nullableTypes.forEach { checkCoverageService?.deepCover(it) }

            if (nonNullableTypes.size == 1) {
                val nonNullableType = nonNullableTypes[0]

                // any and unknown case
                if (isPossiblyNullableType(nonNullableType, context)) {
                    return next(nonNullableType)
                }

                return renderResolvedNullable(nonNullableType, true, next)
            }
        }

        return null
    }

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = Unit
}
